### PR TITLE
feat: identify internal function invocations in traces

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

Introduces `--decode-internal` flag for `forge test`, `cast run` and `cast call --trace` which enables decoding of internal functions in traces

## Example

Example trace of random Uniswap V3 swap:
image

## Solution

To determine when we are jumping in/out of functions we are using source map `Jump` key. However, it is not really reliable, especially after optimizations. Almost in all cases there are mismatches between number of "in"s and "out"s, so we need additional processing to correctly display subset of functions which are correctly reported.

Main implementation of this tracing is in `DebugTraceIdentifier`: https://github.com/foundry-rs/foundry/blob/216e9da8a28fcc57bcba1c6c4986aa5353472cc5/crates/evm/traces/src/debug/mod.rs

The only issue with this approach is that we are losing data about entire stack of internal functions which were joined before revert

I've used default tracer from revm-inspectors instead of traces collected by `Debugger` to allow easier integration into printing logic. Using it required a small patch to inspectors: https://github.com/paradigmxyz/revm-inspectors/pull/150

This approach is enough to implement flamegraphs in a similar way, and can probably be extended to smarter tracking of stack/memory/calldata to also resolve input and output parameters of internal functions

Printing logic is a bit ugly at the moment

Closes: https://github.com/foundry-rs/foundry/issues/3999 + Closes: https://github.com/foundry-rs/foundry/issues/4351

### CODE REVIEW

**Comment by klkvr** [comment]
Added tracking of inputs and ouputs as well. It is currently not able to decode user-defined types such as structs and enums, tracking those would probably require smarter AST analysis.
image

Currently `--decode-internal` is pretty expensive in terms of memory usage because each step are being tracked for each test. We are only interested in `JUMP`s and `JUMPDEST`s, so it might make sense to add a configuration option for `TracingInspector` to only collect steps with specific opcodes.

File: crates/cli/src/utils/cmd.rs
Comment (mattsse) [inline crates/cli/src/utils/cmd.rs:398]: can be done in one step

File: crates/evm/evm/src/inspectors/stack.rs
Comment (mattsse) [inline crates/evm/evm/src/inspectors/stack.rs:401]: this is a bit hard to follow,
I wonder if we can encapsulate these two bools into an enum `TracingKid` or smth, because debug also implies tracing, right?

**Review by mattsse — CHANGES_REQUESTED** [review]
only briefly skimmed parts of it.

I think this makes sense, I'd appreciate a few more docs, and I'll take a closer look

**Comment by klkvr** [comment]
Added `TraceMode` enum which specifies which kind of traces we want: https://github.com/foundry-rs/foundry/blob/a725b2873752cdea3faea759d8b59e23ec367a23/crates/evm/traces/src/lib.rs#L407-L419

After filtering only `JUMP`s and `JUMPDEST`s memory usage is now better (not OOMing anymore on my machine when running big suite without filters), but still very high, so it's better to run `forge test --decode-internal` with filter args

**Comment by klkvr** [comment]
This is pretty much ready. Memory usage is still very high when running a complex test suite with `--decode-internal`, but with filters it is not more than for debugger. `cast run` and `cast call` are also working fine.

Not sure what to do with printing logic. It is not great and I think it would be better to integrate this directly into `TraceWriter` once we have #8224. wdyt @mattsse @DaniPopes

File: crates/cli/src/utils/cmd.rs
Comment (mattsse) [inline crates/cli/src/utils/cmd.rs:387]: can be unwrap_or_default() ?

File: crates/evm/evm/src/executors/invariant/replay.rs
Comment (mattsse) [inline crates/evm/evm/src/executors/invariant/replay.rs:37]: we could add `TraceMode::None` which I think is more ergonomic than doing Option,

but maybe not so we can have an easy conversation into `TracingInspectorConfig`

File: crates/evm/traces/src/debug/sources.rs
Comment (mattsse) [inline crates/evm/traces/src/debug/sources.rs:24]: can be `Arc<str>` I believe

**Review by mattsse — COMMENTED** [review]
cool, only have nits

this is now blocked by the tracewriter pr?
cc @zerosnacks

File: crates/cli/src/utils/cmd.rs
Comment (DaniPopes) [reply crates/cli/src/utils/cmd.rs:387]: not with await

File: crates/evm/traces/src/debug/mod.rs
Comment (ChinW) [inline crates/evm/traces/src/debug/mod.rs:305]: will the usize cause overflow problem, usize is 64 bits (8 bytes), where the length can be 32 bytes long

Comment (ChinW) [reply crates/evm/traces/src/debug/mod.rs:305]: i am trying this new feature, seeing below error in my end

```
Message:  Uint conversion error: Overflow(256, 10488988339550576416, 18446744073709551615)
Location: crates/evm/traces/src/debug/mod.rs:305

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                ⋮ 14 frames hidden ⋮
  15: core::result::Result<T,E>::expect::h1e62bb9f5f4f9d0f
      at /rustc/d124af4d4a707212dfe7c081/library/core/src/result.rs:1034
  16: ruint::from::<impl ruint::Uint<_,_>>::to::h7b50fa5729cd9809
      at /Users/x/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ruint-1.12.3/src/from.rs:236
       234 │         T: Debug,
       235 │     {
       236 >         self.uint_try_to().expect("Uint conversion error")
       237 │     }
       238 │
  17: foundry_evm_traces::debug::decode_from_memory::hb99e9e33dfb84c71
      at /Users/x/foundry/crates/evm/traces/src/debug/mod.rs:305
       303 │         // For `string` and `bytes` layout is a word with length followed by the data
       304 │         DynSolType::String | DynSolType::Bytes => {
       305 >             let length = U256::from_be_bytes::<32>(first_word.try_into().unwrap()).to::<usize>();
       306 │             let data = memory.get(location + 32..location + 32 + length)?;
       307 │
  18: foundry_evm_traces::debug::try_decode_args_from_step::{{closure}}::{{closure}}::h0912b74d67673459
      at /Users/x/foundry/crates/evm/traces/src/debug/mod.rs:283
       281 │                         (DynSolType::Uint(8), Some(Storage::Memory | Storage::Storage)) => None,
```

Comment (klkvr) [reply crates/evm/traces/src/debug/mod.rs:305]: hi @ChinW thanks for trying this out and reporting!

I think it's safe to assume that memory wouldn't be expanded for more than usize, though this analysis is not perfect and we might expect to find the length while there's actually some other value, which I believe has happened in your case

any chance you could provide a repro for this? also, are you using via-ir here?

File: crates/evm/traces/src/debug/sources.rs
Comment (klkvr) [reply crates/evm/traces/src/debug/sources.rs:24]: yea, but this `Arc<String>` is coming from `compilers::Source` and converting it into `Arc<str>` would be expensive I believe

should we change to to `Arc<str>` here? https://github.com/foundry-rs/compilers/blob/8ffdd72e9febd2c9a72d9363f439ce50b59a0db1/crates/artifacts/solc/src/sources.rs#L25

**Comment by Philogy** [comment]
Been waiting for this feature forever!

What do reverts look like? Is it able to give you the line you reverted on?

What about solidity native functions like `abi.encode`, `require`? I think it'd be really useful if the internal calls were shown with a line number

**Comment by klkvr** [comment]
> What do reverts look like? Is it able to give you the line you reverted on?

Current approach relies on solc source map keys of jump type. For JUMPs we are sometimes provided with info on whether this is a jump in or out of the function, and by reading source code you can determine the name, location, input and output types of the function.

However, with optimizations those source maps are getting messed up and you are getting a lot of mismatched ins and outs.

Currently I've been mostly focusing on correctness of the identification, thus we currently only identify a match if we see an explicit JUMP in and JUMP out of the same function. This currently doesn't really work for REVERTs and RETURNs done in low-level assembly, because there are no JUMPs out, just a frame execution end.

We're still ending up with a stack of potentially correct internal fns in those cases, but when I tested this for some random cases this stack usually contained some invalid data which I wouldn't want to display.

So currently there is definitely space for improvement of identification, likely through more "guessing"-approach relying on multiple factors and guided by AST, source maps and bytecode analysis.

> What about solidity native functions like `abi.encode`, `require`? I think it'd be really useful if the internal calls were shown with a line number

While I was reading source maps I've seen that solc marks JUMPs into `abi.encode`/`abi.decode` as jumps in, so it should be possible to identify those in the future. Though this is mostly suited for user-defined internal fns at the moment.

`require` is not treated by solc as a function, though `REVERT` instruction source mapping is usually pointing to the `require` source code, so those should be possible to identify as well.

IMO all of this is basically a better/more readable UX for the debugger, which can already be used to check the exact line of code where the revert occured

**Review by mattsse — APPROVED** [review]
lgtm

pending @DaniPopes

File: crates/evm/traces/src/debug/sources.rs
Comment (klkvr) [reply crates/evm/traces/src/debug/sources.rs:85]: yeah, it is a workaround for https://github.com/foundry-rs/compilers/pull/165

can be removed once we merge it

File: crates/evm/evm/src/executors/trace.rs
Comment (klkvr) [reply crates/evm/evm/src/executors/trace.rs:23]: wouldn't such helpers be too specific? we have only two such branches and they're a bit differen because only one of them accounts for verbosity

File: crates/evm/traces/src/debug/sources.rs
Comment (klkvr) [reply crates/evm/traces/src/debug/sources.rs]: you mean converting it to `Vec<(String, usize, usize)>`?

**Comment by klkvr** [comment]
Updated `--decode-internal` flag to accept a regex similar to debug. On large suites `--decode-internal` easily results in OOM, so I think it's better to restrict it's usage in such way

File: crates/evm/traces/src/debug/sources.rs
Comment (DaniPopes) [reply crates/evm/traces/src/debug/sources.rs]: yes

**Comment by DaniPopes** [comment]
Yeah same problem as in the debugger caused by memory snapshots

File: crates/evm/evm/src/executors/trace.rs
Comment (DaniPopes) [reply crates/evm/evm/src/executors/trace.rs:23]: Better to move the logic in the same place

**Comment by DaniPopes** [comment]
Maybe we can disable memory decoding by default to avoid the memory consumption issue?

**Comment by klkvr** [comment]
> Maybe we can disable memory decoding by default to avoid the memory consumption issue?

Yeah, I though about disabling memory tracking if more than one test matched filters. Though not sure how to make this intuitive

Should it be two separate flags, one of which does not require the test function filter?

File: crates/evm/evm/src/executors/trace.rs
Comment (klkvr) [reply crates/evm/evm/src/executors/trace.rs]: added several builder methods to `TraceMode` 062d55044e7be47038bf73907cb91665231571ad

**Comment by klkvr** [comment]
Updated `forge test --decode-internal` flag to only accept test function parameter optionally:
```
--decode-internal [<TEST_FUNCTION>]
  Whether to identify internal functions in traces.
  
  If no argument is passed to this flag, it will trace internal functions scope and decode stack parameters, but parameters stored in memory (such as bytes or arrays) will not be
  decoded.
  
  To decode memory parameters, you should pass an argument with a test function name, similarly to --debug and --match-test.
  
  If more than one test matches your specified criteria, you must add additional filters until only one test is found (see --match-contract and --match-path).
```

File: crates/forge/bin/cmd/test/mod.rs
Comment (DaniPopes) [inline crates/forge/bin/cmd/test/mod.rs:319]: Can we have an enum just for this? Full, Simple, No

**Review by DaniPopes — APPROVED** [review]

File: crates/forge/bin/cmd/test/mod.rs
Comment (klkvr) [reply crates/forge/bin/cmd/test/mod.rs]: added in 92ece42

**Review by mattsse — APPROVED** [review]

**Comment by Philogy** [comment]
Hey @klkvr, great work on this PR!

Just wanted to follow up on your comment about source maps getting messed up by the optimization step: do you mean that they're actually broken and it'd be useful to open an issue in solc or just that optimization fundamentally obfuscates the origin of an opcode as they could be reduced to fewer operations?

**Comment by klkvr** [comment]
@Philogy it's hard to tell, I've definitely seen situations in which source maps would point to completely unrelated chunck of code for some of the instructions after optimizations. However, I wouldn't be surprised if this has a reasonable explanation related to how inlining works internally. Sourcemaps are documented very briefly so it's hard to tell how they should behave and what we should consider a bug, and whether we can trust them after certain number of optimizations at all

**Comment by m-waqas88** [comment]
Version: `forge 0.2.0 (41d4e54 2024-09-17T00:21:10.066464000Z)`

Screenshot 2024-09-18 at 12 31 12 AM

Hey guys, really appreciate your efforts. I am actually facing an issue, when using `--decode-internal` flag, the value of arguments shown in the trace is actually wrong, as you can see I have logged the actual values pass as arguments. `epoch` is the first parameter and `account` is the second one. Console logging is showing correct values but in call trace (as function inputs) this value is incorrect.

Notes:
- In foundry.toml --via-ir is set to true (I cannot disable it otherwise it will throw stack too deep error)
- `forge test --mp VaultWY.t.sol --mt test_WY_withdrawAndClaim -vvv --decode-internal --fork-url <fork-url>`

@klkvr

**Comment by m-waqas88** [comment]
> Version: `forge 0.2.0 (41d4e54 2024-09-17T00:21:10.066464000Z)`
> 
> Screenshot 2024-09-18 at 12 31 12 AM
> 
> Hey guys, really appreciate your efforts. I am actually facing an issue, when using `--decode-internal` flag, the value of arguments shown in the trace is actually wrong, as you can see I have logged the actual values pass as arguments. `epoch` is the first parameter and `account` is the second one. Console logging is showing correct values but in call trace (as function inputs) this value is incorrect.
> 
> Notes:
> 
> * In foundry.toml --via-ir is set to true (I cannot disable it otherwise it will throw stack too deep error)
> * `forge test --mp VaultWY.t.sol --mt test_WY_withdrawAndClaim -vvv --decode-internal --fork-url <fork-url>`
> 
> @klkvr

Ok I got it. first parameter `1335805085125855948948476260874682295582732906387` is the decimal representation of the address --> `waqas`, and the second parameter is the hex representation of zero `0`.

**Pull request merged**
